lean_source/06_Abstract_Algebra/source_01_Structures.lean

```diff
@@ -0,0 +1,503 @@
+import algebra.big_operators.ring
+import data.real.basic
+
+/- TEXT:
+
+.. _section_structures:
+
+Structures
+----------
+
+In the broadest sense of the term, a *structure* is a specification
+of a collection of data, possibly with contraints that the
+data is required to satisfy.
+An *instance* of the structure is a particular bundle of data satisfying
+the constraints. For example, we can specify that a point is
+a tuple of three real numbers:
+BOTH: -/
+-- QUOTE:
+@[ext] structure point := (x : ℝ) (y : ℝ) (z : ℝ)
+-- QUOTE.
+
+/- TEXT:
+The ``@[ext]`` annotation tells Lean to automatically generate theorems
+that can be used to prove that two instances of a structure are equal
+when their components are equal.
+EXAMPLES: -/
+-- QUOTE:
+#check point.ext
+
+example (a b : point) (hx : a.x = b.x) (hy : a.y = b.y) (hz : a.z = b.z) :
+  a = b :=
+begin
+  ext,
+  repeat { assumption }
+end
+-- QUOTE.
+
+/- TEXT:
+We can then define particular instances of the ``point`` structure.
+Lean provides multiple ways of doing that.
+EXAMPLES: -/
+-- QUOTE:
+def my_point1 : point :=
+{ x := 2,
+  y := -1,
+  z := 4 }
+
+def my_point2 :=
+{ point .
+  x := 2,
+  y := -1,
+  z := 4 }
+
+def my_point3 : point := ⟨2, -1, 4⟩
+
+def my_point4 := point.mk 2 (-1) 4
+-- QUOTE.
+
+/- TEXT:
+In the first two examples, the fields of the structure are named
+explicitly.
+In the first case, because Lean knows that the expected type of
+``my_point1`` is a ``point``, you can start the definition by
+writing ``{! !}``. Clicking on the light bulb
+that appears nearby in VS Code will then
+give you the option of inserting a template definition
+with the field names listed for you.
+
+By default, the *constructor* for the ``point`` structure is named
+``point.mk``. You can specify a different name if you want.
+EXAMPLES: -/
+-- QUOTE:
+structure point' := build :: (x : ℝ) (y : ℝ) (z : ℝ)
+
+#check point'.build 2 (-1) 4
+-- QUOTE.
+
+/- TEXT:
+The next two examples show how to define functions
+on structures using pattern matching.
```

Comment:
Why not starting with the much easier version without pattern matching?
```lean
def point.add (p q : point) : point :=
{ x := p.x + q.x,
  y := p.y + q.y,
  z := p.z + q.z }
```
This would allow to introduce the projection notation that is referred to without explanation below.